lib/ansible/modules/crypto/get_certificate.py

```diff
@@ -210,45 +275,81 @@ def main():
 
             cert = ctx.wrap_socket(sock, server_hostname=host).getpeercert(True)
             cert = DER_cert_to_PEM_cert(cert)
-            x509 = crypto.load_certificate(crypto.FILETYPE_PEM, cert)
         except Exception as e:
             module.fail_json(msg="Failed to get cert from port with error: {0}".format(e))
 
     else:
         try:
             cert = get_server_certificate((host, port), ca_certs=ca_cert)
-            x509 = crypto.load_certificate(crypto.FILETYPE_PEM, cert)
         except Exception as e:
             module.fail_json(msg="Failed to get cert from port with error: {0}".format(e))
 
     result['cert'] = cert
-    result['subject'] = {}
-    for component in x509.get_subject().get_components():
-        result['subject'][component[0]] = component[1]
-
-    result['expired'] = x509.has_expired()
-
-    result['extensions'] = []
-    extension_count = x509.get_extension_count()
-    for index in range(0, extension_count):
-        extension = x509.get_extension(index)
-        result['extensions'].append({
-            'critical': extension.get_critical(),
-            'asn1_data': extension.get_data(),
-            'name': extension.get_short_name(),
-        })
-
-    result['issuer'] = {}
-    for component in x509.get_issuer().get_components():
-        result['issuer'][component[0]] = component[1]
-
-    result['not_after'] = x509.get_notAfter()
-    result['not_before'] = x509.get_notBefore()
-
-    result['serial_number'] = x509.get_serial_number()
-    result['signature_algorithm'] = x509.get_signature_algorithm()
 
-    result['version'] = x509.get_version()
+    if backend == 'pyopenssl':
+        x509 = crypto.load_certificate(crypto.FILETYPE_PEM, cert)
+        result['subject'] = {}
+        for component in x509.get_subject().get_components():
+            result['subject'][component[0]] = component[1]
+
+        result['expired'] = x509.has_expired()
+
+        result['extensions'] = []
+        extension_count = x509.get_extension_count()
+        for index in range(0, extension_count):
+            extension = x509.get_extension(index)
+            result['extensions'].append({
+                'critical': extension.get_critical(),
+                'asn1_data': extension.get_data(),
+                'name': extension.get_short_name(),
+            })
+
+        result['issuer'] = {}
+        for component in x509.get_issuer().get_components():
+            result['issuer'][component[0]] = component[1]
+
+        result['not_after'] = x509.get_notAfter()
+        result['not_before'] = x509.get_notBefore()
+
+        result['serial_number'] = x509.get_serial_number()
+        result['signature_algorithm'] = x509.get_signature_algorithm()
+
+        result['version'] = x509.get_version()
+
+    elif backend == 'cryptography':
+        x509 = cryptography.x509.load_pem_x509_certificate(to_bytes(cert), cryptography_backend())
+        result['subject'] = {}
```

Comment:
Technically subject components need to be ordered, so a dict wouldn't work well.

However, I get the impression that the purpose of this module is informative rather than actionable (particularly since it doesn't return thumbprint/public key/etc), in which case I would agree that it doesn't matter if the results are ordered, since it would be an abuse of this module in the first place to be _doing things_ based on subject DN, when it comes to certificate validation/chain building/etc.

Reply:
True indeed. Probably the best solution would be to return a list of `(type, value)` pairs. Not very handy to work with, though :)

My assumption is also that the result of this module (except the certificate itself) is mostly informational: one can quickly get some information from it without processing the certificate further. Like check whether the common name is `My Company` instead of `Whatever Else`.